Add partially defined functions

BjoernPetersen opened the pull request:
Adds support for partially defined functions to the ABS language.  
This includes the addition of `map`, `filter`, `foldl` and `foldr` functions to the StdLib.

For information on the usage of the added features, please consult the updated abs-docs.

Note: this pull request should probably be squashed.

rudi commented:
Thanks!  Will start reviewing this week.

rudi commented on frontend/src/abs/frontend/typechecker/Typing.jrag:
```diff
@@ -386,6 +391,11 @@ aspect Typing {
     public interface HasTypes {
       java.util.List<Type> getTypes();
     }
+
```
In case we remove explicit type parameters at call sites, this will need a `getType()` method as well:
`AnonymousFunctionDecl.getType() = this.getPureExp().getType();`

Similar for `NamedParFnAppParam`: look up the function by name, call `getType` on the Decl, return UnknownType if not found.

rudi commented on frontend/src/abs/lang/abslang.abs:
```diff
@@ -50,6 +50,10 @@ export Inv, Pre, Post;
 export Annotation, TypeAnnotation, LocationType, Far, Near, Somewhere, Infer;
 export COG, Plain, Final, Atomic;
 export HTTPName, HTTPCallable;
+export Expansion, ExpansionCall;
```
Do we need to export these?  Is there anything a modeler (user of the language) can do with these annotations?

BjoernPetersen replied:
The annotations could be useful for static analysis and IDE support, but modelers should, indeed, not use them. However, not exporting the annotation types leads to type-checking errors for the generated annotations. Is there any way I don't know about to circumvent that?

rudi replied:
If we can keep them purely an implementation detail, you can use a similar approach to Lars, see file `ASTPreProcessor.java` line 177, and grep for `FUNCTIONSELECTOR` to see where he uses it

BjoernPetersen replied:
I see. The problem with using this approach is that it just "marks" a function as an accessor function. The Expansion and ExpansionCall annotations also refer to each other (e.g. `[ExpansionCall: list[1]]` refers to a function with the `[Expansion: 1]` annotation). As I see it, there is no way of doing that with Lars' approach and still be able to identify the annotation types.

rudi commented:
Sorry for the incorrect comment locations in the review!  I mostly left them above the code I was commenting on, the tool expects them below, so you don't see the code I'm commenting on in the thread above :(

BjoernPetersen commented:
Thanks for the quick review! I'll give it a closer look tommorow

rudi commented on frontend/src/abs/lang/abslang.abs:
```diff
@@ -595,6 +598,26 @@ data HTTPCallableAnnotation = HTTPCallable;
+    Nil => Nil;
+    Cons(x, xs) => Cons(f(x), map(xs));
+};
+
```
Hmm, we already have a function `map` (the `Map` constructor function).  At the moment ABS doesn't have function overloading, so I'd expect a compile-time error here.

Incidentally, this last commit breaks the Java backend (the created file `Map.java` which defines the Map datatype is empty).  I think it's unlikely that this commit actually introduces the bug rather than just exposing it, but didn't investigate further yet which commit actually introduces the breakage.

rudi replied:
On second thought, since normal and partial functions are syntactically distinct, it might be ok to have them in separate namespaces.

BjoernPetersen replied:
I also assume the namespaces won't collide. I'll look into the problem with the Java-Backend though

BjoernPetersen replied:
The problem with the Java backend should be fixed in c7a24e8cdd30cf7367701172da4cc399c8ed3eec .

rudi commented on frontend/src/abs/frontend/pardef/ParFnAppParam.jrag:
```diff
@@ -0,0 +1,173 @@
+    * Stores the IDs for all pure expressions. Keys are the pretty-printed pure expressions.
+    */
+    static Map<String, String> PureExp.pureExpIds = new HashMap<>(256);
+
```
Could you run this by @rm2340 ? I could potentially see this static cache having an impact on the work he's doing in dynamic / multiple product generation.

BjoernPetersen replied:
I realized that assigning IDs to pure expressions was the result of a premature optimization regarding anonymous functions. If an anonymous function is used multiple times, it should probably just be declared as a regular, named function.

BjoernPetersen commented on abs-docs/asciidoc/functional.adoc:
```diff
@@ -605,3 +634,150 @@ NOTE: The ABS standard library contains some special functions that cannot be
+This is not used for code generation, but helps error reporting.
+
+[source]
+----
```
I agree, I'll remove implementation-specific details from the manual

BjoernPetersen commented on frontend/src/abs/frontend/typechecker/Typing.jrag:
```diff
@@ -93,6 +93,10 @@ aspect Typing {
     eq TypeSynDecl.getType() { 
         return getValue().getType(); 
     }
+
```
PartialFunctionDecl instances actually don't vanish anymore, neither do ParFnApp instances which are children of a PartialFunctionDecl.
I guess UnknownType would be the more sensible choice here, although type checking is effectively disabled for instances of both PartialFunctionDecl and ParFnApp.

I'll just try and see that the type-checker throws at me if I change it.

rudi replied:
Not having PartialFunctionDecl vanish would be good news!  Then you can round-trip pretty-print them (print the PartialFunctionDecl and don't pretty-print the generated functions; check what Lars has been doing with the generated datatype accessor functions).

BjoernPetersen replied:
PartialFunctionDecls are printed already. Not printing generated expansions would be trivial, but I don't think we should do that, because then we should also print the generated FnApp as the original ParFnApp, which would be cumbersome.

BjoernPetersen commented on frontend/src/abs/frontend/pardef/PartialFunctionExpander.jadd:
```diff
@@ -0,0 +1,399 @@
+
+            PartialFunctionDecl function = getPartialFunction(fnApp.getName());
+            if (function == null) {
+                throw new PardefModellingException(fnApp, ErrorMessage.FUNCTION_NOT_RESOLVABLE, fnApp.getName());
```
This should be fixed now (73645cec7180dfe549e98c4c29a882d6f0ccdd00)

BjoernPetersen commented:
@rudi I have implemented the type parameter inference for pardefs and rebased the branch on the current abstools/abstools master branch. I believe this pull request is ready to merge now.

rudi commented:
Great work, thanks!  To my delight, anonymous functions support local variables, as in the example below.

```
module Test;
{
    Int value = 3;
    List<Int> l = filter((String x) => x > value)(list[1, 2, 3, 4, 5]);
    println(toString(l));
}
```

One small niggle - this model reports an error at a wrong location:

```
$ ~/Source/abstools/frontend/bin/bash/absc -erlang "foo.abs" 
foo.abs:1:0:Comparison expression with incomparable types, ABS.StdLib.String and ABS.StdLib.Int.
module Test;
^
foo.abs:4:4:Type ABS.StdLib.List<ABS.StdLib.Int> does not match declared type ABS.StdLib.List<ABS.StdLib.String>.
    List<Int> l = filter((String x) => x > value)(list[1, 2, 3, 4, 5]);
---^

Compilation failed with syntax errors.
```

Note the 1:0 location in the first error message — this should probably in Line 4, Column 39 or thereabouts.

I’m niggling about this since Emacs uses this information to highlight the source file :)

Rudi

> On 26 Nov 2017, at 15:19, Björn Petersen <notifications@github.com> wrote:
> 
> @rudi <https://github.com/rudi> I have implemented the type parameter inference for pardefs and rebased the branch on the current abstools/abstools master branch. I believe this pull request is ready to merge now.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/abstools/abstools/pull/174#issuecomment-347011753>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAAUflVq-jE9bpWRXNMIw58Wb-xAQBnKks5s6XOMgaJpZM4QTas7>.
>

BjoernPetersen commented:
I hope the error messages are all fixed now

rudi commented:
There were some invalid test cases, I'll fix them.  Many thanks again for your work!
